### PR TITLE
Get XRC20 transfers without specifying contract address 83

### PR DESCRIPTION
**Pull request by lzxm160**
work on #83 #84 

**1.byContractAddress**
query{
xrc20{
  byContractAddress(
    address:"io14j96vg9pkx28htpgt2jx0tf3v9etpg4j9h384m",
numPerPage:10,page:1
  ){
    exist,
    xrc20{
      hash,
from,
to,
quantity,
timestamp,
    },
    count,
  }
}
}
**2.byRecipientAddress**
query{
xrc20{
  byRecipientAddress(
    address:"io1rxwu5wd24y2nf5t836uu7yky85p5f0pvquj59m",
numPerPage:10,page:1
  ){
    exist,
    xrc20{
      hash,
from,
to,
quantity,
timestamp,
    },
    count,
  }
}
}
**3.byPage**
query{
xrc20{
  byPage(
numPerPage:10,page:1
  ){
    exist,
    xrc20{
      hash,
from,
to,
quantity,
timestamp,
    },
    count,
  }
}
}

### CODE REVIEW

File: graphql/resolver.go [hunk 1]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get numPerPage")
+	}
+	page, err := getIntArg(argsMap, "page")
+	if err != nil {
+		return errors.Wrap(err, "failed to get page")
+	}
+	xrc20InfoList, err := r.AP.GetXrc20(address, uint64(numPerPage), uint64(page))
+	switch {
+	case errors.Cause(err) == indexprotocol.ErrNotExist:
```
Comment (lizhefeng) [inline graphql/resolver.go]: Shouldn't set Exist to false and return in this case?

File: graphql/resolver.go [hunk 2]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+			Hash:      c.Hash,
+			Timestamp: c.Timestamp,
+			From:      c.From,
+			To:        c.To,
+			Quantity:  c.Quantity,
+		})
+	}
+
+	output := &Xrc20List{Exist: true, Count: len(xrc20InfoRet), Xrc20: xrc20InfoRet}
+	if len(xrc20InfoRet) == 0 {
```
Comment (lizhefeng) [inline graphql/resolver.go]: If we get here, is it possible that the length is 0?

File: graphql/resolver.go [hunk 3]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get numPerPage")
+	}
+	page, err := getIntArg(argsMap, "page")
+	if err != nil {
+		return errors.Wrap(err, "failed to get page")
+	}
+	xrc20InfoList, err := r.AP.GetXrc20ByRecipient(address, uint64(numPerPage), uint64(page))
+	switch {
+	case errors.Cause(err) == indexprotocol.ErrNotExist:
```
Comment (lizhefeng) [inline graphql/resolver.go]: Same here.

File: graphql/resolver.go [hunk 4]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+			Hash:      c.Hash,
+			Timestamp: c.Timestamp,
+			From:      c.From,
+			To:        c.To,
+			Quantity:  c.Quantity,
+		})
+	}
+
+	output := &Xrc20List{Exist: true, Count: len(xrc20InfoRet), Xrc20: xrc20InfoRet}
+	if len(xrc20InfoRet) == 0 {
```
Comment (lizhefeng) [inline graphql/resolver.go]: Ditto.

File: graphql/resolver.go [hunk 5]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get numPerPage")
+	}
+	page, err := getIntArg(argsMap, "page")
+	if err != nil {
+		return errors.Wrap(err, "failed to get page")
+	}
+	xrc20InfoList, err := r.AP.GetXrc20ByPage(uint64(numPerPage), uint64(page))
+	switch {
+	case errors.Cause(err) == indexprotocol.ErrNotExist:
```
Comment (lizhefeng) [inline graphql/resolver.go]: Same here.

File: graphql/resolver.go [hunk 6]
```diff
@@ -451,7 +447,122 @@ func (r *queryResolver) getActionsByDates(ctx context.Context, actionResponse *A
...
+			Hash:      c.Hash,
+			Timestamp: c.Timestamp,
+			From:      c.From,
+			To:        c.To,
+			Quantity:  c.Quantity,
+		})
+	}
+
+	output := &Xrc20List{Exist: true, Count: len(xrc20InfoRet), Xrc20: xrc20InfoRet}
+	if len(xrc20InfoRet) == 0 {
```
Comment (lizhefeng) [inline graphql/resolver.go]: Ditto.

File: graphql/resolver.go [hunk 7]
```diff
@@ -680,7 +791,13 @@ func getIntArg(argsMap map[string]*ast.Value, argName string) (int, error) {
 	}
 	return intVal, nil
 }
-
+func getStringArg(argsMap map[string]*ast.Value, argName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (lzxm160) [reply graphql/resolver.go]: sure

**Pull request merged**